File: src/MyServiceBus/BusFactoryBuilder.cs
using System;
using System.Collections.Generic;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;

namespace MyServiceBus;

public class BusFactoryBuilder : IBusFactory
{
    readonly IBusFactory inner;
    ILoggerFactory? loggerFactory;
    readonly IList<Action<IServiceCollection>> serviceActions = new List<Action<IServiceCollection>>();

    public BusFactoryBuilder(IBusFactory inner)
    {
        this.inner = inner;
    }

    public BusFactoryBuilder WithLoggerFactory(ILoggerFactory factory)
    {
        loggerFactory = factory;
        return this;
    }

    public BusFactoryBuilder ConfigureServices(Action<IServiceCollection> configure)
    {
        serviceActions.Add(configure);
        return this;
    }

    public IMessageBus Create<T>(Action<T>? configure = null) where T : IBusFactoryConfigurator, new()
    {
        var cfg = new T();
        configure?.Invoke(cfg);

        var services = new ServiceCollection();
        foreach (var action in serviceActions)
            action(services);
        if (loggerFactory != null)
            services.AddSingleton(loggerFactory);

        cfg.Configure(services);
        var provider = services.BuildServiceProvider();

        foreach (var action in provider.GetServices<IPostBuildAction>())
            action.Execute(provider);

        return provider.GetRequiredService<IMessageBus>();
    }
}

File: src/Java/myservicebus/src/main/java/com/myservicebus/BusFactoryBuilder.java
package com.myservicebus;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

import com.myservicebus.di.ServiceCollection;
import com.myservicebus.di.ServiceProvider;
import com.myservicebus.logging.LoggerFactory;

public class BusFactoryBuilder implements BusFactory {
    private final BusFactory inner;
    private LoggerFactory loggerFactory;
    private final List<Consumer<ServiceCollection>> serviceActions = new ArrayList<>();

    BusFactoryBuilder(BusFactory inner) {
        this.inner = inner;
    }

    public BusFactoryBuilder withLoggerFactory(LoggerFactory factory) {
        this.loggerFactory = factory;
        return this;
    }

    public BusFactoryBuilder configureServices(Consumer<ServiceCollection> configure) {
        this.serviceActions.add(configure);
        return this;
    }

    @Override
    public <T extends BusFactoryConfigurator> MessageBus create(Class<T> configuratorClass, Consumer<T> configure) {
        try {
            T cfg = configuratorClass.getDeclaredConstructor().newInstance();
            if (configure != null) {
                configure.accept(cfg);
            }

            ServiceCollection services = ServiceCollection.create();
            for (Consumer<ServiceCollection> action : serviceActions) {
                action.accept(services);
            }
            if (loggerFactory != null) {
                services.addSingleton(LoggerFactory.class, sp -> () -> loggerFactory);
            }

            cfg.configure(services);
            ServiceProvider provider = services.buildServiceProvider();
            return provider.getService(MessageBus.class);
        } catch (Exception e) {
            throw new RuntimeException("Unable to create configurator", e);
        }
    }
}
